Crashlytics/Crashlytics/Models/FIRCLSOnDemandModel.m

```diff
@@ -0,0 +1,193 @@
+  _baseExponent = baseExponent;
+  _stepDuration = stepDuration;
+
+  _recordedOnDemandExceptionCount = 0;
+
+  _buckets = [NSMutableArray array];
+  _lastUpdated = [NSDate timeIntervalSinceReferenceDate];
+  _currentStep = -1;
+
+  for (int i = 0; i < uploadRate; i++) {
+    [_buckets addObject:[NSNumber numberWithInt:-1]];
+  }
+  sharedOnDemandModel = self;
+  return sharedOnDemandModel;
+}
+
+/*
+ * Called from FIRCrashlytics whenever the on-demand record exception method is called.
+ */
+- (BOOL)recordOnDemandExceptionIfQuota:(FIRExceptionModel *)exceptionModel
+             withDataCollectionEnabled:(BOOL)dataCollectionEnabled
+            usingExistingReportManager:(FIRCLSExistingReportManager *)existingReportManager {
+  // Record the exception model into a new report if there is unused on-demand quota. Otherwise,
+  // log the occurence but drop the event.
+  @synchronized(self) {
+    if (![self isQueueFull]) {
+      FIRCLSDataCollectionToken *dataCollectionToken = [FIRCLSDataCollectionToken validToken];
+      NSString *activeReportPath = FIRCLSExceptionRecordOnDemandModel(
+          exceptionModel, self.recordedOnDemandExceptionCount, self.droppedOnDemandExceptionCount);
+
+      // Only submit an exception report if data collection is enabled. Otherwise, the report
+      // is stored until send or delete unsent reports is called.
+      double uploadDelay = [self calculateUploadDelay:YES];
+      if (dataCollectionEnabled) {
+        if (activeReportPath) {
+          // dispatch after calculateUploadDelay then update bucket with time of upload
+          [existingReportManager handleOnDemandReportUpload:activeReportPath
+                                        dataCollectionToken:dataCollectionToken
+                                                  withDelay:uploadDelay
+                                                   asUrgent:YES];
+          FIRCLSInfoLog(@"Submitted an on-demand exception with delay %.20f", uploadDelay);
+          [self updateBucketsAndCountWithDelay:uploadDelay];
+          return YES;  // Recorded and submitted the exception.
+        } else {
+          FIRCLSInfoLog(@"Error recording on-demand exception");
+          return NO;  // Something went wrong when recording the exception, so we don't have a valid
+                      // path.
+        }
+      }
+      FIRCLSInfoLog(@"Submitted an on-demand exception with delay %d", uploadDelay);
+      [self updateBucketsAndCountWithDelay:uploadDelay];
+      return YES;  // Recorded exception but did not submit since data collection is disabled.
+    } else {
+      FIRCLSInfoLog(@"No available on-demand quota. Dropping report.");
+      self.droppedOnDemandExceptionCount++;
+      [self calculateUploadDelay:YES];
+      return NO;  // Didn't record or submit then exception because no quota was available.
+    }
+  }
+}
+
+// Helper methods
+
+- (double)calculateUploadDelay:(BOOL)shouldUpdate {
+  // TODO: should return 0 if queue is empty?
```

Comment:
If the queue is empty it should still calculate the delay just like a "NotFull" queue.

Reply:
It defaults to a six second delay though - I think we want to have it upload immediately if there haven't been any events uploaded before